### ISSUE TITLE
Buggy vmap

### ISSUE THREAD

**Issue by Atcold** [pytorch/pytorch#100410]
### 🐛 Describe the bug

Spurious singleton dimension.

```python
import torch
from torch import optim, nn
from torch.nn import functional; nn.f = functional
from torch.func import vmap, jacrev


class VariationalAutoencoder(nn.Module):
    def __init__(self, w=200, D=20):
        super().__init__()
        self.D = D    # latent size
        NL = nn.ReLU  # non-linearity
        self.encoder = nn.Sequential(
            nn.Linear(2, w), NL(),
            nn.Linear(w, w), NL(),
            nn.Linear(w, 2 * D),
        )
        self.decoder = nn.Sequential(
            nn.Linear(D, w), NL(),
            nn.Linear(w, w), NL(),
            nn.Linear(w, 2)
        )

    def reparameterise(self, μ, log_σ2):
        if self.training:
            σ = log_σ2.mul(0.5).exp_()
            z = torch.randn(self.D, device=σ.device)
            return z.mul(σ).add_(μ)
        return μ

    def forward(self, y):
        h = self.encoder(y)
        μ_lσ2 = h.view(-1, 2, self.D)
        μ      = μ_lσ2[:, 0, :]
        log_σ2 = μ_lσ2[:, 1, :]
        ζ = self.reparameterise(μ, log_σ2)
        ỹ = self.decoder(ζ)
        if self.training:
            return ỹ, μ, log_σ2
        return ỹ


def vmap_test(net):
    n = lambda b: nn.f.normalize(b, dim=-1)
    u = lambda ȳ: n(net(ȳ) - ȳ)
    with torch.no_grad():
        u_Ȳ = u(Ȳ)
        J_u_Ȳ = vmap(jacrev(u))(Ȳ)
    print(u_Ȳ.size(), J_u_Ȳ.size())


Ȳ = torch.randn(961, 2)

vae = VariationalAutoencoder()
vae.eval()

vmap_test(vae)
```

```
$ python vmap_bug.py 
torch.Size([961, 2]) torch.Size([961, 1, 2, 2])
```

Also, `nn.f.normalize(dim=-1)` should be the default. `dim=1` is just **bad**.

### Versions

Latest.

cc @zou3519 @Chillee @samdow @soumith @kshitij12345 @janeyx99

**Comment by kshitij12345** [comment]
I don't think the output under vmap is buggy.

`μ_lσ2 = h.view(-1, 2, self.D)` is interesting as it will add a singleton dimension, in case `h` was not batched but if it is batched then it would be no-op.

```python
>>> import torch
# assuming `self.D == 2`
# Unbatched
>>> x = torch.randn(2, 2)
>>> x.view(-1, 2, 2).shape
torch.Size([1, 2, 2])
# Batched
>>> x = torch.randn(3, 2, 2)
>>> x.view(-1, 2, 2).shape
torch.Size([3, 2, 2])
```

With this in mind, calling `u` on unbatched sample `u(Ȳ[0])` returns `torch.Size([1, 2])`. So running this under vmap (semantically: for-loop over batch dim followed by stack) correctly preserves this shape and returns `torch.Size([961, 1, 2])`.

**Comment by Atcold** [comment]
I'm not sure I follow @hrbigelow.
`B = 961` in my example.

**Comment by zou3519** [comment]
Closing as resolved

**Comment by Atcold** [comment]
The solution proposed by @zou3519 is to replace
```python
μ_lσ2 = h.view(-1, 2, self.D)
μ      = μ_lσ2[:, 0, :]
log_σ2 = μ_lσ2[:, 1, :]
```
with
```python
μ_lσ2 = h.unflatten(-1, (2, self.D))
μ      = μ_lσ2.select(-2, 0)
log_σ2 = μ_lσ2.select(-2, 1)
```
so that a singleton is not created when using individual samples.
